document-store/src/main/java/org/hypertrace/core/documentstore/postgres/utils/PostgresUtils.java

```diff
@@ -450,25 +455,31 @@ public static String prepareParsedNonCompositeFilter(
       case "NOT_IN":
       case "NOT IN":
         // NOTE: Pl. refer this in non-parsed expression for limitation of this filter
+        // In order to make the behaviour same as for mongo, the not in operator would match only if
+        // the lhs
+        // and rhs have no intersection at all
+        // NOTE: This doesn't work in case the lhs is a function
         sqlOperator = " NOT IN ";
         isMultiValued = true;
-        isInOP = true;
         filterString
             .append(" IS NULL OR")
-            .append(" NOT ")
+            .append(" NOT (")
             .append(
                 prepareFilterStringForInOperator(
-                    preparedExpression, (Iterable<Object>) value, paramsBuilder));
-        break;
+                    preparedExpression, (Iterable<Object>) value, paramsBuilder))
+            .append(")");
+        return filterString.toString();
       case "IN":
+        // In order to make the behaviour same as for mongo, the in operator would match if the lhs
+        // and rhs have any intersection at all
         // NOTE: Pl. refer this in non-parsed expression for limitation of this filter
+        // NOTE: This doesn't work in case the lhs is a function
```

Comment:
```suggestion
        // In order to make the behaviour same as for Mongo, the "INI" operator would match if the 
        // LHS and RHS have any intersection (i.e. non-empty intersection)
        // NOTE: Pl. refer this in non-parsed expression for limitation of this filter
        // NOTE: This doesn't work in case the LHS is a function
```